File: Lua/fractionalization.lua
-- © 2019 Aaron Sami Abassi
-- Licensed under the Academic Free License version 3.0
local function greatest_common_divisor(a, b)
    -- Euclidean Algorithm
    local dividend, divisor, remainder
    if a > b then
        dividend = a
        divisor = b
    else
        dividend = a
        divisor = b
    end
    remainder = dividend % divisor
    while remainder > 0 do
        dividend = divisor
        divisor = remainder
        remainder = dividend % divisor
    end
    return divisor
end
local function least_common_multiple(a, b)
    local divisor = greatest_common_divisor(a, b)
    if a > b then 
        return a / divisor * b
    end
    return b / divisor * a
end
local function reduce(fraction)
    if fraction.numerator == 0 then
        return {numerator = 0, denominator = 1}
    end
    local divisor = greatest_common_divisor(fraction.numerator, fraction.denominator)
    return {
        numerator = fraction.numerator / divisor, 
        denominator = fraction.denominator / divisor
    }
end
local function fast_add(base, relative)
    return {
        numerator = base.numerator * relative.denominator + relative.numerator * base.denominator,
        denominator = base.denominator * relative.denominator
    }
end
local function fast_subtract(base, relative)
    return {
        numerator = base.numerator * relative.denominator - relative.numerator * base.denominator,
        denominator = base.denominator * relative.denominator
    }
end
local function fast_multiply(base, relative)
    return {
        numerator = base.numerator * relative.numerator,
        denominator = base.denominator * relative.denominator
    }
end
local function fast_divide(base, relative)
    return {
        numerator = base.numerator * relative.denominator,
        denominator = base.denominator * relative.numerator
    }
end
local function reducing_add(base, relative)
    local multiple = least_common_multiple(base.denominator, relative.denominator)
    local result = {}
    result.numerator = base.numerator * (multiple / base.denominator) + relative.numerator * (multiple / relative.denominator)
    result.denominator = multiple
    return reduce(result)
end
local function reducing_subtract(base, relative)
    local multiple = least_common_multiple(base.denominator, relative.denominator)
    local result = {}
    result.numerator = base.numerator * (multiple / base.denominator) - relative.numerator * (multiple / relative.denominator)
    result.denominator = multiple
    return reduce(result)
end
local function reducing_multiply(base, relative)
    local bnrd_divisor, bdrn_divisor, result
    if base.numerator > 0 then
        bnrd_divisor = greatest_common_divisor(base.numerator, relative.denominator)
    else
        bnrd_divisor = 1
    end
    if relative.numerator > 0 then
        bdrn_divisor = greatest_common_divisor(base.denominator, relative.numerator)
    else
        bdrn_divisor = 1
    end
    return {
        numerator = (base.numerator / bnrd_divisor) * (relative.numerator / bdrn_divisor),
        denominator = (base.denominator / bdrn_divisor) * (relative.denominator / bnrd_divisor)
    }
end
local function reducing_divide(base, relative)
    local bnrn_divisor = greatest_common_divisor(base.numerator, relative.numerator)
    local bdrd_divisor = greatest_common_divisor(base.denominator, relative.denominator)
    return {
        numerator = (base.numerator / bnrn_divisor) * (relative.denominator / bdrd_divisor),
        denominator = (base.denominator / bdrd_divisor) * (relative.numerator / bnrn_divisor)
    }
end
local function fast_lesser(base, relative)
    return base.numerator * relative.denominator < relative.numerator * base.denominator
end
local function fast_greater(base, relative)
    return base.numerator * relative.denominator > relative.numerator * base.denominator
end
local function fast_equal(base, relative)
    return base.numerator * relative.denominator == relative.numerator * base.denominator
end
local function fast_not_greater(base, relative)
    return base.numerator * relative.denominator <= relative.numerator * base.denominator
end
local function fast_not_lesser(base, relative)
    return base.numerator * relative.denominator >= relative.numerator * base.denominator
end
local function fast_not_equal(base, relative)
    return base.numerator * relative.denominator ~= relative.numerator * base.denominator
end
local function reducing_lesser(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) < relative.numerator * (factor / relative.denominator)
end
local function reducing_greater(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) > relative.numerator * (factor / relative.denominator)
end
local function reducing_equal(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) == relative.numerator * (factor / relative.denominator)
end
local function reducing_not_greater(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) <= relative.numerator * (factor / relative.denominator)
end
local function reducing_not_lesser(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) >= relative.numerator * (factor / relative.denominator)
end
local function reducing_not_equal(base, relative)
    local factor = least_common_multiple(base.denominator, relative.denominator)
    return base.numerator * (factor / base.denominator) ~= relative.numerator * (factor / relative.denominator)
end
local fast_arithmetic = {
    add = fast_add,
    subtract = fast_subtract,
    multiply = fast_multiply,
    divide = fast_divide
}
local reducing_arithmetic = {
    add = reducing_add,
    subtract = reducing_subtract,
    multiply = reducing_multiply,
    divide = reducing_divide
}
local fast_relation = {
    lesser = fast_lesser,
    greater = fast_greater,
    equal = fast_equal,
    not_greater = fast_not_greater,
    not_lesser = fast_not_lesser,
    not_equal = fast_not_equal
}
local reducing_relation = {
    lesser = reducing_lesser,
    greater = reducing_greater,
    equal = reducing_equal,
    not_greater = reducing_not_greater,
    not_lesser = reducing_not_lesser,
    not_equal = reducing_not_equal
}
local fast_operation = {
    arithmetic = fast_arithmetic,
    relation = fast_relation
}
local reducing_operation = {
    arithmetic = reducing_arithmetic,
    relation = reducing_relation
}
return {
    greatest_common_divisor = greatest_common_divisor,
    least_common_multiple = least_common_multiple,
    fast_add = fast_add,
    fast_subtract = fast_subtract,
    fast_multiply = fast_multiply,
    fast_divide = fast_divide,
    reducing_add = reducing_add,
    reducing_subtract = reducing_subtract,
    reducing_multiply = reducing_multiply,
    reducing_divide = reducing_divide,
    fast_lesser = fast_lesser,
    fast_greater = fast_greater,
    fast_equal = fast_equal,
    fast_not_greater = fast_not_greater,
    fast_not_lesser = fast_not_lesser,
    fast_not_equal = fast_not_equal,
    reducing_lesser = reducing_lesser,
    reducing_greater = reducing_greater,
    reducing_equal = reducing_equal,
    reducing_not_greater = reducing_not_greater,
    reducing_not_lesser = reducing_not_lesser,
    reducing_not_equal = reducing_not_equal,
    fast_arithmetic = fast_arithmetic,
    reducing_arithmetic = reducing_arithmetic,
    fast_relation = fast_relation,
    reducing_relation = reducing_relation,
    fast_operation = fast_operation,
    reducing_operation = reducing_operation
}

File: Go/fractionalization.go
// © 2019 Aaron Sami Abassi
// Licensed under the Academic Free License version 3.0
package fractionalization
type Fractional struct {
    Numerator, Denominator uint
}
type ArithmeticalAbstract func(Fractional, Fractional) Fractional
type RelationalAbstract func(Fractional, Fractional) bool
type Arithmetical struct {
    Add, Subtract, Multiply, Divide ArithmeticalAbstract
}
type Relational struct {
    Lesser, Greater, Equal, NotGreater, NotLesser, NotEqual RelationalAbstract
}
type Operational struct {
    Arithmetic Arithmetical
    Relation Relational
}
func GreatestCommonDivisor(a uint, b uint) uint {
    // Euclidean Algorithm
    var (dividend, divisor, remainder uint)
    if a > b {
        dividend = a
        divisor = b
    } else {
        dividend = b
        divisor = a
    }
    remainder = dividend % divisor
    for remainder > 0 {
        dividend = divisor
        divisor = remainder
        remainder = dividend % divisor
    }
    return divisor
}
func LeastCommonMultiple(a uint, b uint) uint {
    divisor := GreatestCommonDivisor(a, b)
    if a > b {
        return a / divisor * b
    }
    return b / divisor * a
}
func Reduce(fraction Fractional) Fractional {
    var divisor uint
    if fraction.Numerator == 0 {
        return Fractional{0, 1}
    }
    divisor = GreatestCommonDivisor(fraction.Numerator, fraction.Denominator)
    return Fractional{fraction.Numerator / divisor, fraction.Denominator / divisor}
}
func FastAdd(base Fractional, relative Fractional) Fractional {
    return Fractional{
        base.Numerator * relative.Denominator + relative.Numerator * base.Denominator,
        base.Denominator * relative.Denominator,
    }
}
func FastSubtract(base Fractional, relative Fractional) Fractional {
    return Fractional{
        base.Numerator * relative.Denominator - relative.Numerator * base.Denominator,
        base.Denominator * relative.Denominator,
    }
}
func FastMultiply(base Fractional, relative Fractional) Fractional {
    return Fractional{
        base.Numerator * relative.Numerator,
        base.Denominator * relative.Denominator,
    }
}
func FastDivide(base Fractional, relative Fractional) Fractional {
    return Fractional{
        base.Numerator * relative.Denominator,
        relative.Numerator * base.Denominator,
    }
}
func ReducingAdd(base Fractional, relative Fractional) Fractional {
    multiple := LeastCommonMultiple(base.Denominator, relative.Denominator)
    result := Fractional{
        base.Numerator * (multiple / base.Denominator) + relative.Numerator * (multiple / relative.Denominator),
        multiple,
    }
    return Reduce(result);
}
func ReducingSubtract(base Fractional, relative Fractional) Fractional {
    multiple := LeastCommonMultiple(base.Denominator, relative.Denominator)
    result := Fractional{
        base.Numerator * (multiple / base.Denominator) - relative.Numerator * (multiple / relative.Denominator),
        multiple,
    }
    return Reduce(result);
}
func ReducingMultiply(base Fractional, relative Fractional) Fractional {
    var (bnrd_divisor, bdrn_divisor uint)
    if base.Numerator > 0 {
        bnrd_divisor = GreatestCommonDivisor(base.Numerator, relative.Denominator)
    } else {
        bnrd_divisor = 1
    }
    if relative.Numerator > 0 {
        bdrn_divisor = GreatestCommonDivisor(base.Denominator, relative.Numerator)
    } else {
        bdrn_divisor = 1
    }
    return Fractional{
        (base.Numerator / bnrd_divisor) * (relative.Numerator / bdrn_divisor),
        (base.Denominator / bdrn_divisor) * (relative.Denominator / bnrd_divisor),
    }
}
func ReducingDivide(base Fractional, relative Fractional) Fractional {
    bnrn_divisor := GreatestCommonDivisor(base.Numerator, relative.Numerator)
    bdrd_divisor := GreatestCommonDivisor(base.Denominator, relative.Denominator)
    return Fractional{
        (base.Numerator / bnrn_divisor) * (relative.Denominator / bdrd_divisor),
        (base.Denominator / bdrd_divisor) * (relative.Numerator / bnrn_divisor),
    }
}
func FastLesser(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator < relative.Numerator * base.Denominator
}
func FastGreater(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator > relative.Numerator * base.Denominator
}
func FastEqual(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator == relative.Numerator * base.Denominator
}
func FastNotGreater(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator <= relative.Numerator * base.Denominator
}
func FastNotLesser(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator >= relative.Numerator * base.Denominator
}
func FastNotEqual(base Fractional, relative Fractional) bool {
    return base.Numerator * relative.Denominator != relative.Numerator * base.Denominator
}
func ReducingLesser(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) < relative.Numerator * (factor / relative.Denominator)
        
}
func ReducingGreater(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) > relative.Numerator * (factor / relative.Denominator)
        
}
func ReducingEqual(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) == relative.Numerator * (factor / relative.Denominator)
        
}
func ReducingNotGreater(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) <= relative.Numerator * (factor / relative.Denominator)
        
}
func ReducingNotLesser(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) >= relative.Numerator * (factor / relative.Denominator)
        
}
func ReducingNotEqual(base Fractional, relative Fractional) bool {
    factor := LeastCommonMultiple(base.Numerator, relative.Denominator)
    return base.Numerator * (factor / base.Denominator) != relative.Numerator * (factor / relative.Denominator)
        
}
func FastArithmetic() Arithmetical {
    return Arithmetical{FastAdd, FastSubtract, FastMultiply, FastDivide}
}
func ReducingArithmetic() Arithmetical {
    return Arithmetical{ReducingAdd, ReducingSubtract, ReducingMultiply, ReducingDivide}
}
func FastRelation() Relational {
    return Relational{FastLesser, FastGreater, FastEqual, FastNotGreater, FastNotLesser, FastNotEqual}
}
func ReducingRelation() Relational {
    return Relational{ReducingLesser, ReducingGreater, ReducingEqual, ReducingNotGreater, ReducingNotLesser, ReducingNotEqual}
}
func FastOperation() Operational {
    return Operational{FastArithmetic(), FastRelation()}
}
func ReducingOperation() Operational {
    return Operational{ReducingArithmetic(), ReducingRelation()}
}

